[1단계 - 로또 구현] 노랑(노주희) 미션 제출합니다.

yesjuhee opened the pull request:
리뷰 잘 부탁드립니다 😄 

추가로 질문 드리고 싶은 것이 있어서 정리해봤습니다.

**1. MVC 구조**
미션 구현에 MVC 구조를 사용해서 작성했는데, 올바르게 작성된 구조인지 궁금합니다!
또한 다른 크루들의 코드를 봤을 때 **서비스**를 작성해서 구현하는 경우를 봤는데 서비스는 어떤 역할을 하는건가요?
서치를 해봤는데 MVC 아키텍처에서 서비스가 어떤 역할을 하는 것인지 정확히 이해가 되지 않아서 질문드립니다.

**2. 테스트 작성 시점**
이번 미션이 테스트에 대한 미션이다보니 페어랑 구현하면서 가능하면 모든 기능에 대해서 테스트를 작성하려고 해보았는데요 그 과정에서 몇가지 궁금한 점이 있었습니다.

예를 들어 코드를 아래와 같이 작성한 상태라고 가정해보겠습니다.
```java
public class LottoNumbers {
    private final List<Integer> numbers;

    public LottoNumbers() {
        this.numbers = generateLottoNumbers();
    }

    private List<Integer> generateLottoNumbers() {
        // 테스트 해보고 싶은 로직
    }
}    
```
`generateLottoNumbers()` 메서드를 작성한 후, 해당 메서드가 복잡한 로직을 가지고 있기 때문에 정상적으로 작동하는지 테스트를 해보고 싶었습니다. 그런데 아직 `LottoNumbers`를 생성하는 로직 까지만 작성하고 꺼내오는 로직은 작성하지 않았기 때문에 게터를 추가하지 않는 한 테스트가 불가능했습니다. 또한 일반적으로 게터를 지양해야한다고 알고 있기 때문에 결국 직접 실행해서 기능을 테스트하거나, 해당 값을 꺼내오는 로직을 작성할 때까지 테스트를 미루는 방식으로 진행했습니다.

혹시 이 문제의 해결책이나 대체할 수 있는 방법이 있을까요?

https://prolog.techcourse.co.kr/studylogs/3861

yesjuhee committed:
docs: 리드미 작성

yesjuhee committed:
feat : 구입금액 안내멘트를 출력한다.

yesjuhee committed:
feat : 구입금액을 입력한다.

yesjuhee committed:
feat : 양의 정수가 아닌 입력은 예외 처리한다.

yesjuhee committed:
feat : 1,000원으로 나누어떨어지지 않는 입력은 예외 처리한다.

yesjuhee committed:
test : 구매 금액 입력 테스트

yesjuhee committed:
feat : 6개의 랜덤한 숫자를 구입 개수만큼 생성한다.

yesjuhee committed:
feat : 구매 결과를 출력한다.

yesjuhee committed:
feat : 생성된 번호를 화면에 출력한다.

yesjuhee committed:
feat : 번호 입력 안내멘트를 출력한다.

yesjuhee committed:
feat : 쉼표로 나눴을 때 6개가 아닌 경우 예외 처리한다.

yesjuhee committed:
feat : 양의 정수가 아닌 입력은 예외 처리한다.

yesjuhee committed:
feat : 1~45 범위가 아닌 입력은 예외 처리한다.

yesjuhee committed:
feat : 중복된 번호를 입력할 경우 예외 처리한다.

yesjuhee committed:
feat : 보너스 번호 입력 안내멘트를 출력한다.

yesjuhee committed:
feat : 보너스 번호 입력을 예외 처리한다.

yesjuhee committed:
feat : 당첨 번호와 중복된 번호를 입력할 경우 예외 처리한다.

yesjuhee committed:
보너스 번호를 입력받는다.

yesjuhee committed:
feat : 각 로또별로 당첨번호와 일치하는 숫자 개수를 카운팅한다.

yesjuhee committed:
feat : 당첨 결과를 출력한다.

yesjuhee committed:
feat : 총 당첨금을 계산한다.

yesjuhee committed:
feat : 계산된 수익률을 출력한다

yesjuhee committed:
refactor : 테스트 형식 수정

jinyoungchoi95 commented on src/main/java/controller/PurchaseController.java:
> 미션 구현에 MVC 구조를 사용해서 작성했는데, 올바르게 작성된 구조인지 궁금합니다!
> 또한 다른 크루들의 코드를 봤을 때 서비스를 작성해서 구현하는 경우를 봤는데 서비스는 어떤 역할을 하는건가요?
> 서치를 해봤는데 MVC 아키텍처에서 서비스가 어떤 역할을 하는 것인지 정확히 이해가 되지 않아서 질문드립니다.

미션에서 mvc에 대한 이야기가 나오는 것은 mvc에서 controller나 더 나아가서 서비스를 쓰라는 것이 아니라 `MVC 패턴 기반으로 리팩터링해 view 패키지의 객체가 domain 패키지 객체에 의존할 수 있지만, domain 패키지의 객체는 view 패키지 객체에 의존하지 않도록 구현한다.`입니다.

즉, Controller를 쓰라는게 아니라 도메인을 잘 설계하고 도메인 안에 화면의 요구사항을 넣지말라는거죠. 1레벨에서는 controller와 service라는 네이밍은 모두 쓸 일이 없고, 각 객체에 적절한 네이밍을 달아주면 됩니다.

yesjuhee replied:
오 그렇군요 명쾌한 답변 감사합니다 💯 
미션 방향 설정에 많은 도움 될 것 같아요~~!!

jinyoungchoi95 commented on src/main/java/model/LottoNumbers.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Random;
+import java.util.Set;
+
+public class LottoNumbers {
```
> generateLottoNumbers() 메서드를 작성한 후, 해당 메서드가 복잡한 로직을 가지고 있기 때문에 정상적으로 작동하는지 테스트를 해보고 싶었습니다. 그런데 아직 LottoNumbers를 생성하는 로직 까지만 작성하고 꺼내오는 로직은 작성하지 않았기 때문에 게터를 추가하지 않는 한 테스트가 불가능했습니다. 또한 일반적으로 게터를 지양해야한다고 알고 있기 때문에 결국 직접 실행해서 기능을 테스트하거나, 해당 값을 꺼내오는 로직을 작성할 때까지 테스트를 미루는 방식으로 진행했습니다.

`Assertions.isEqualTo`가 어떻게 동작하고, 클래스의 equals 재정의를 하였을때와 하지 않았을때 `Assertions.isEqualTo`가 어떤 결과를 내는지 확인해볼까요?

yesjuhee replied:
헉 완전 이해했습니다
클래스 자체를 넣어서 테스트가 가능하군요...!
완전 유레카네요 😮

jinyoungchoi95 commented on README.md:
> JUnit5와 AssertJ의 주요 차이점은 무엇인가?
> JUnit5는 actual, expect를 한 번에 받고, AssertJ는 따로 받는다?

한 번에 받고 따로 받았을 때 어떤 동작이 달라지나요? assertJ를 사용하는 이유는 무엇때문인가요?

yesjuhee replied:
JUnit은 테스트 할 값과 실제 값을 하나의 매개변수의 인수로 받기 때문에 매개변수가 바뀌는 등 실수를 하기 쉽습니다.
AssertJ는 인자를 하나만 갖는 메서드를 연결해서 사용하기 때문에 실수가 발생할 확률이 적고, 가독성이 높아져서 사용합니다!

jinyoungchoi95 commented on src/test/java/model/LottoConstantsTest.java:
> AssertJ를 활용하여 특정 조건을 검증하는 테스트 코드를 작성할 수 있는가?
> 넵

실제 코드에서 AssertJ를 어떻게 활용하여 어떤 조건을 사용하여 테스트했나요?

yesjuhee replied:
`assertThat().isEqualTo()`
`assertThat().isTrue()`
`assertThat().isBetween()`
과 같은 형태로 사용했습니다!

jinyoungchoi95 replied:
학습로그를 작성하실때는 나온 질문에 대해서 더 자세하게 작성해주는 것이 이후로도 미션을 진행하고 학습하시는데 많은 도움이 될 것 같아요.

yesjuhee replied:
넵 감사합니다!

jinyoungchoi95 commented on src/main/java/constant/Constants.java:
```diff
@@ -0,0 +1,9 @@
+package constant;
+
+public class Constants {
```
1레벨에서의 미션은 얼마나 도메인 객체를 잘 설계했고 잘 활용하느냐인데요. `Constants`는 도메인객체에서 정보를 벗어나게하여 도메인 응집도를 떨어뜨립니다.

각 정보는 필요한 도메인 객체가 가지고 사용할 수 있도록 수정하면 좋을 것 같아요.

yesjuhee replied:
넵 반영하겠습니다 🫡

yesjuhee replied:
3953478

jinyoungchoi95 commented on src/main/java/model/LottoNumbers.java:
로또 미션에서 가장 중요한 객체인데 `countMatchNumber`나 `bonusMatch`도 테스트가 필요하지 않을까요?

yesjuhee replied:
`LottoNumbers`의 해당 메서드들을 `WinLotto`에서 호출한 후 바로 리턴하는 형식으로 구현을 했습니다.
`WinLotto`의 해당 메서드들에 대한 테스트를 작성한 후 `LottoNumbers`의 해당 메서드들을 테스트 하는 코드는 완전 똑같은 기능을 중복해서 테스트하는 것이라고 판단해서 페어랑 이야기해서 생략했습니다!

다시 생각해보니까 신뢰할 수 있는 코드를 위해서는 추가로 테스트 코드를 작성하는게 좋을 것 같네욥 
반영하겠습니다 🫡 


```java
// (참고용)
public class WinLotto {
   // ...

    public Integer countMatchNumber(LottoNumbers lottoNumbers) {
        return this.lottoNumbers.countMatchNumber(lottoNumbers);
    }

    public Boolean bonusMatch(LottoNumbers lottoNumbers) {
        return lottoNumbers.bonusMatch(this.bonusNumber);
    }
}

yesjuhee replied:
25511fd

jinyoungchoi95 commented on src/main/java/model/WinLotto.java:
```diff
@@ -0,0 +1,29 @@
+    private final LottoNumbers lottoNumbers;
+    private final Integer bonusNumber;
+
+    public WinLotto(List<Integer> winNumbers, Integer bonusNumber) {
```
생성시에 `bonusNumber`에 100이 들어오면 비정상 객체가 생성될 수 있겠네요 :)

yesjuhee replied:
입력 받은 보너스 번호 값에 대한 범위 체크는 `WinLottoView`에서 입력을 받은 직후에 `validateBound()` 메서드를 통해서 하고 있습니다.
뷰에서 파싱을 모두 하고 `Integer` 혹은 `List<Integer>`의 타입으로 모델에 입력 값을 넘겨주도록 구현을 했기 때문에 파싱하는 과정에서 유효성 검사까지 모두 뷰에 작성했습니다.

validation 메서드들은 전부 도메인에 위치도록 수정하는게 좋을까요?

yesjuhee replied:
리팩터링을 하다가 위 질문에 대해서 스스로 나름의 답을 내리게 되었습니다.
비즈니스 로직과 직접 연관이 있는 메서드만 도메인에 위치하도록 수정했습니다!

도메인에 위치한 검증 로직
- 구매 금액이 로또 1장 가격으로 나누어 떨어지는지 검증하는 메서드
- 로또 번호, 보너스 번호의 범위가 1~45 사이인지 검증하는 메서드
- 입력한 로또 번호의 개수가 6개인지 검증하는 메서드

뷰에 위치한 검증 로직
- 입력한 숫자가 양의 정수인지 검증하는 메서드

c8d5815

e2fbfd7

yesjuhee replied:
그..리고 validation 위치를 바꾸면서 지금 구조에서 `validateBound()` 메서드가 불가피하게 `LottoNumbers`와 `WinLotto`에 중복되어 위치하게 되었는데요.

이에 대해 고민해보다가 숫자의 범위가 1~45인지 판단하는 것은 결국 "하나의 숫자"에 대한 책임이기 때문에 `LottoNumber`라는 별도의 객체를 이용해서 원시값 포장을 하고 해당 클래스에 `validationBound()` 메서드를 위치하도록 리팩터링하는 방향을 생각했습니다.

그런데 리팩터링을 하려고 보니까 수정해야 할 부분이 너무 많아서 step2에서 한 번 시도해보겠습니다ㅎㅎ

jinyoungchoi95 commented on src/main/java/model/Purchase.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Purchase {
```
`구입`이라는 객체의 네이밍을 고민해보면 좋을 것 같아요. 코드보다는 실제를 생각해보면 "`구입`이 `로또`를 사고 구매 갯수를 계산한다"라는 말이 어색하지 않나요?

yesjuhee replied:
`PurchasedLottos`로 좀 더 명확하게 표현하는게 좋을 것 같네요.
감사합니다 🫡

yesjuhee replied:
0b34289

jinyoungchoi95 commented on src/main/java/model/Result.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.List;
+import java.util.Map.Entry;
+
+public class Result {
```
"결과"라는 이름보다는 객체를 더 잘 표현해줄 수 있는 이름으로 변경해보면 좋을 것 같네요.

yesjuhee replied:
`LottoStatistics`로 변경하면 좋을 것 같네요 감사합니다 👍

yesjuhee replied:
4f8b097

jinyoungchoi95 commented on src/main/java/model/Result.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.Map.Entry;
+
+public class Result {
+    private final EnumMap<WinLottoInfo, Integer> result = new EnumMap<>(WinLottoInfo.class);
```
`EnumMap`을 사용한 이유가 있을까요? 어디에 장점이 있나요?

yesjuhee replied:
EnumMap은 Enum의 ordinal을 인덱스로 사용해서 데이터를 저장합니다.
따라서 해시 테이블을 사용하지 않기 때문에 HashMap과 비교했을 때 성능이 더 좋습니다!
또한 ordinal을 이용하기 때문에 순서를 보장하지 않는 HashMap과 비교해서 Enum선언순서대로 순서를 유지할 수 있습니다.

jinyoungchoi95 commented on src/main/java/model/Result.java:
```diff
@@ -0,0 +1,35 @@
+
+    private Integer calculateTotalPrize() {
+        Integer sum = 0;
+        for (Entry<WinLottoInfo, Integer> winLottoInfoIntegerEntry : result.entrySet()) {
```
`winLottoInfoIntegerEntry`도 있는 그대로의 네이밍을 붙여서 작성하기보다는 조금 더 잘 표현할 수 있게 변경해보면 좋을 것 같아요.

yesjuhee replied:
`statisticsEntry`가 적절할 것 같습니다. 감사합니다 👍

yesjuhee replied:
377609c

jinyoungchoi95 commented on src/test/java/constant/WinLottoInfoTest.java:
```diff
@@ -0,0 +1,68 @@
+    List<WinLottoInfo> resultList;
+    List<WinLottoInfo> winLottoInfo = Arrays.stream(WinLottoInfo.values()).toList();
+
+    @BeforeEach
+    void setUp() {
```
junit5는 기본 Lifecycle이 pre-test이기 때문에 굳이 `@BeforeEach`로 분리하여 생성할 필요는 없습니다.

https://junit.org/junit5/docs/current/user-guide/#writing-tests-test-instance-lifecycle

yesjuhee replied:
오 감사합니다 반영하겠습니다 🫡

yesjuhee replied:
95549df

yesjuhee replied:
95549df

jinyoungchoi95 commented on src/main/java/model/Purchase.java:
```diff
@@ -0,0 +1,31 @@
+    private final Integer purchaseAmount;
+
+    public Purchase(Integer purchaseAmount) {
+        Integer purchaseCount = purchaseAmount / Constants.LOTTO_PRICE;
```
`Integer`가 아니라 `int`여도 되지 않을까요?

yesjuhee replied:
확장성을 고려해서 항상 원시타입대신 래퍼클래스를 써왔는데요...!
혹시 현업에서는 래퍼클래스의 메서드를 사용하지 않는 경우 원시타입을 쓰나요? 궁금합니다!

yesjuhee replied:
신경쓰지 않아서 몰랐는데 `Integer`를 쓰는 경우 IDE에서 경고 메시지가 뜨네요
그래서 그냥 수정했습니다!

c9cb563

jinyoungchoi95 replied:
래퍼클래스의 단점을 생각해보면 좋을 것 같아요. 굳이 래퍼클래스를 선호하지는 않습니다.

jinyoungchoi95 reviewed (changes requested):
안녕하세요 노랑 😄

로또 미션의 리뷰어를 맡은 오리입니다.

첫 미션 잘 구현해주셨네요. 다음 단계 넘어가기 전 같이 이야기하면 좋을 내용 코멘트 남겨두었으니 확인부탁드려요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

yesjuhee committed:
refactor : 상수 클래스 분리

- 각 상수를 책임지는 역할을 하는 클래스로 위치를 변경함
- 여러 클래스에서 필요한 상수의 경우, 가장 책임이 큰 클래스 하나에 상수를 위치시킨 후 접근 제어자를 public으로 설정함

yesjuhee committed:
refactor : 테스트 코드 추가 작성

- LottoNumbers 클래스의 countMatchNumber() 테스트 코드 작성
- LottoNumbers 클래스의 bonusMatch() 테스트 코드 작성

yesjuhee committed:
refactor : 일부 validate 메서드의 위치를 뷰에서 도메인으로 변경

- 구매 금액이 로또 1장 가격으로 나누어 떨어지는지 검증하는 메서드
- 로또 번호, 보너스 번호의 범위가 1~45 사이인지 검증하는 메서드
- 입력한 로또 번호의 개수가 6개인지 검증하는 메서드

yesjuhee committed:
refactor : validate 메서드를 좀 더 연관이 높은 도메인으로 위치 변경, 테스트 코드 수정

yesjuhee committed:
chore : Purchase를 PurchasedLottos로 이름 변경

yesjuhee committed:
chore : Result를 LottoStatistics로 이름 변경

yesjuhee committed:
chore : 클래스 이름, 변수명 변경

yesjuhee committed:
refactor : 불필요한 테스트 코드 삭제

yesjuhee committed:
chore : warning 문제 해결

yesjuhee commented:
양질의 리뷰 덕분에 미션을 200% 깨닫게 된 것 같습니다 감사합니다 🙇 🙇‍♂️ 🙇‍♀️

jinyoungchoi95 commented on src/main/java/constant/ErrorMessage.java:
```diff
@@ -0,0 +1,10 @@
+    public static final String POSITIVE_NUMBER_EXCEPTION = "양의 정수를 입력해주세요.";
+    public static final String DIVIDABLE_EXCEPTION = "1000의 배수를 입력해주세요.";
+    public static final String NUMBER_COUNT_EXCEPTION = "6개의 숫자를 입력해주세요.";
+    public static final String NUMBER_BOUND_EXCEPTION = "1~45 사이의 숫자를 입력해주세요.";
+    public static final String NUMBER_DUPLICATE_EXCEPTION = "중복이 아닌 숫자를 입력해주세요";
+    public static final String BONUS_NUMBER_DUPLICATE_EXCEPTION = "당첨 번호와 중복이 아닌 숫자를 입력해주세요";
```
`ErrorMessage`라는 것이 혼자서 무언가의 역할을 하고 있는 객체로 보이지 않는데 필요한 클래스가 맞을까요? 미션내에서 객체지향에 대해서 고민해보시고 "상수 클래스" "상수 패키지"라는 것이 객체지향에서 필요한지 고민해보시면 좋을 것 같아요.

jinyoungchoi95 commented on src/main/java/constant/WinLottoInfo.java:
```diff
@@ -0,0 +1,48 @@
+import model.LottoNumbers;
+import model.WinLotto;
+
+public enum WinLottoInfo {
```
등수를 측정하는 도메인 클래스 아닌가요? constant 패키지에 있는 이유가 있을까요?

jinyoungchoi95 commented on src/main/java/constant/WinLottoInfo.java:
```diff
@@ -0,0 +1,48 @@
+import model.WinLotto;
+
+public enum WinLottoInfo {
+    FIRST(6, false, 2000000000),
+    SECOND(5, true, 30000000),
+    THIRD(5, false, 1500000),
+    FOURTH(4, false, 50000),
+    FIFTH(3, false, 5000),
```
```suggestion
    FIRST(6, false, 2_000_000_000),
    SECOND(5, true, 30_000_000),
    THIRD(5, false, 1_500_000),
    FOURTH(4, false, 50_000),
    FIFTH(3, false, 5_000),
```
숫자는 가독성을 위해 _로 구분지을 수 있습니다.

jinyoungchoi95 commented on src/main/java/constant/WinLottoInfo.java:
```diff
@@ -0,0 +1,48 @@
+        this.price = price;
+    }
+
+    public static WinLottoInfo result(LottoNumbers purchasedLotto, WinLotto winLotto) {
```
메소드는 동작을 나타내는 동사여야합니다. result는 결과라는 명사인데 메소드의 의미를 잘 표현해주고 있지 않네요.

jinyoungchoi95 commented on src/main/java/controller/LottoStatisticsController.java:
```diff
@@ -0,0 +1,16 @@
+import view.LottoStatisticsView;
+
+public class LottoStatisticsController {
+    public LottoStatisticsView lottoStatisticsView = new LottoStatisticsView();
```
전반적으로 접근제어자와 불변이 불규칙하게 사용되고 있는 것이 보이는데 모두 기준을 맞추어 설정하면 좋을 것 같네요.

jinyoungchoi95 commented on src/main/java/model/LottoNumbers.java:
```diff
@@ -0,0 +1,69 @@
+        this.numbers = generateLottoNumbers();
+    }
+
+    public LottoNumbers(List<Integer> numbers) {
```
메소드의 순서가 public, private 섞여있는데 메소드를 배치하는 규칙이 있을까요?

jinyoungchoi95 commented on src/test/java/model/LottoConstantsTest.java:
```diff
@@ -0,0 +1,31 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class LottoConstantsTest {
```
어떤 것을 위한 테스트일까요? `LottoNumbers`에 대한 테스트라면 `LottoNumbersTest`로 옮겨야할 것 같네요.

jinyoungchoi95 commented on src/test/java/model/LottoNumbersTest.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    @Test
+    @DisplayName("숫자 범위 판별 테스트")
```
현재 숫자의 threshold는 1~45이기 때문에 46만 테스트하는 것이 아닌 0이 걸리는 경우도 테스트하면 조금 더 안정적인 테스트케이스가 되겠네요.

jinyoungchoi95 commented on src/main/java/constant/WinLottoInfo.java:
```diff
@@ -0,0 +1,48 @@
+    NONE(2, false, 0);
+
+    private final Integer matchNumberCount;
+    private final Boolean matchBonusNumber;
```
사용하지 않고 있네요.

jinyoungchoi95 commented on src/main/java/constant/WinLottoInfo.java:
```diff
@@ -0,0 +1,48 @@
+            return THIRD;
+        }
+        List<WinLottoInfo> filteredWinLottoInfo = Arrays.stream(WinLottoInfo.values())
+                .filter((winLottoInfo) -> winLottoInfo.matchNumberCount == matchNumberCount)
```
<img width="1049" alt="스크린샷 2025-02-15 오후 4 36 57" src="https://github.com/user-attachments/assets/6b3e0c3e-a28a-40a0-be37-4fcdac04d9f6" />

인텔리제이가 왜 경고문구를 보내고 있을까요?

jinyoungchoi95 commented on src/main/java/constant/WinLottoInfo.java:
[클린코드 2. 의미있는 이름](https://github.com/Yooii-Studios/Clean-Code/blob/master/Chapter%2002%20-%20%EC%9D%98%EB%AF%B8%20%EC%9E%88%EB%8A%94%20%EC%9D%B4%EB%A6%84.md)

클래스 이름에 Info, Data와 같은 불용어를 붙이지 않는다.

jinyoungchoi95 reviewed (approved):
안녕하세요 노랑 😄

피드백 잘 반영해주셨네요. 몇가지 코멘트 남겨둔 것은 다음 단계에 함께 반영부탁드려요.
미션 코멘트 알림을 받기 힘들어 궁금하신점은 다음 단계 pr에 함께 올려주시면 다음 단계에서 같이 확인하겠습니다.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

yesjuhee commented:
@jinyoungchoi95 
감사합니다! 확인해서 다음 PR에 반영하겠습니다!
